### PR TITLE
feature/direct-to-pdf-BZ-4508

### PR DESCRIPTION
**Pull request by tiandavis**
## Summary - [BZ-4508](https://thirdiron.atlassian.net/browse/BZ-4508)

Add Direct To PDF Link when available - Primo and Summon

## Description

Tasks:
- Add Direct To PDF Link when available - Primo (priority) && Summon (DONE)
Primo (DONE)
Summon (DONE)
Unit and Acceptance Tests (DONE)
Cross Browser Tests (Chrome, FF, Safari, IE11, Edge) (DONE)
- Summon: Refactor template to position "Article PDF" the first entry (DONE)
- Primo: Refactor template to align "Download PDF" with native "Available Online" Link (DONE)
- Override the default link text shown (DONE)
- On by default, but can disable with a config flag (DONE)
- Remove browzineEnabled check for CoverImages (DONE)
- Always update the Journal Cover when available, unless the Journal Cover is the default image (DONE)

**Screenshots**

Primo

image

Summon

image

## Deploy Precautions

- None

### CODE REVIEW

File: tests/acceptance/browzine-primo-adapter-test.js [hunk 1]
```diff
@@ -169,11 +170,19 @@ describe("BrowZine Primo Adapter >", function() {
...
       expect(template).toBeDefined();
-      expect(template.text().trim()).toEqual("View Issue Contents");
+
+      expect(template.text().trim()).toContain("View Issue Contents");
+      expect(template.text().trim()).toContain("Download Now");
+
       expect(template.find("a.browzine-web-link").attr("href")).toEqual("https://browzine.com/libraries/XXX/journals/18126/issues/7764583?showArticleInContext=doi:10.1136/bmj.h2575");
       expect(template.find("a.browzine-web-link").attr("target")).toEqual("_blank");
-      expect(template.find("img.browzine-book-icon").attr("src")).toEqual("https://assets.thirdiron.com/images/integrations/browzine_open_book_icon.png");
+      expect(template.find("img.browzine-book-icon").attr("src")).toEqual("https://assets.thirdiron.com/images/integrations/browzine-open-book-icon.png");
```
Comment (karlbecker) [inline tests/acceptance/browzine-primo-adapter-test.js]: dash-er-ized >> snake_case ? 😄

File: tests/unit/models/primo.js [hunk 2]
```diff
@@ -617,7 +618,120 @@ describe("Primo Model >", function() {
...
+    });
+
+    it("should return true when a default coverImageUrl is returned by the API", function() {
+      var coverImageUrl = "https://assets.thirdiron.com/images/covers/default.png";
+      expect(primo.isDefaultCoverImage(coverImageUrl)).toEqual(true);
+    });
+  });
+
+  describe("primo model getDirectToPDFUrl method >", function() {
+    it("should not include a fullTextFile property in the BrowZine API response for a journal", function() {
```
Comment (karlbecker) [inline tests/unit/models/primo.js]: I'd suggest a little rearrangement of the test here to match the given / when / then sort of format that is [popular with Cucumber](https://docs.cucumber.io/gherkin/reference/#steps) and generally [a BDD-style approach](https://martinfowler.com/bliki/GivenWhenThen.html).

How about something sort of like this?
```
describe("primo model getDirectToPDFUrl method >", function() {
  describe("given an API response for a journal", ...
    var scope = { ... issn: ["0096-6762" ...
    it("should not include a fullTextFile property", ...
```

I was at first unclear about where the test setup was and what the fixture was used for.  

After reading the test again, I can see the fixture is all about a journal response - and yup, journal responses don't have `fullTextFile` properties! 😊 

Typically we have fixtures setup in a `describe` setup, since the `describe` is usually a `given` or `when`.

What do you think of this approach, @tiandavis ?  Also, we could address this after this PR is merged if we'd like to get this deployed for John's demo tomorrow.

Thanks!

File: tests/unit/models/primo.js [hunk 3]
```diff
@@ -617,7 +618,120 @@ describe("Primo Model >", function() {
...
+      };
+
+      var data = primo.getData(journalResponse);
+
+      expect(data).toBeDefined();
+
+      expect(primo.getDirectToPDFUrl(scope, data)).toBeNull();
+    });
+
+    it("should include a fullTextFile property in the BrowZine API response for an article", function() {
```
Comment (karlbecker) [inline tests/unit/models/primo.js]: Same thing here - 
describe("primo model getDirectToPDFUrl method >", function() {
  describe("given an API response for a journal", ... 
  );
  describe("given an API response for an article", ...
    var scope = { ... doi: ["10.1136/bmj.h2575" ...
    it("should include a fullTextFile property", ...
```

Thoughts?

File: tests/unit/models/primo.js [hunk 4]
```diff
@@ -617,7 +618,120 @@ describe("Primo Model >", function() {
...
+    it("should show direct to pdf link when configuration property is undefined or null", function() {
+      expect(primo.showDirectToPDFLink()).toEqual(true);
+    });
+
+    it("should show direct to pdf link when configuration property is true", function() {
+      browzine.primoArticlePDFDownloadLinkEnabled = true;
+      expect(primo.showDirectToPDFLink()).toEqual(true);
+    });
+
+    it("should hide direct to pdf link when configuration property is false", function() {
```
Comment (karlbecker) [inline tests/unit/models/primo.js]: 👍 to this unit test.

Do we want an acceptance-level test that checks to ensure there's no element that appears when this configuration property is false, too?  Similar to [this assertion](https://github.com/thirdiron/browzine-discovery-service-adapters/pull/45/files#diff-e918fc280a5c48d536bacb6a76a7de33R730), but we'd have a test and setup that makes sure the fully-output data truly excludes the PDF stuff.  

Basically, a test to ensure that `showDirectToPDFLink()` has an impact on the final HTML.

File: src/summon/browzine-summon-adapter.js [hunk 5]
```diff
@@ -150,10 +150,60 @@ browzine.summon = (function() {
...
+    if(typeof config === "undefined" || config === null || config === true) {
+      enableShowDirectToPDFLink = true;
+    }
+
+    return enableShowDirectToPDFLink;
+  };
+
+  function directToPDFTemplate(directToPDFUrl) {
+    var pdfIcon = "https://s3.amazonaws.com/thirdiron-assets/images/integrations/browzine-pdf-download-icon.png";
+    var articlePDFDownloadWording = browzine.summonArticlePDFDownloadWording || "Article PDF";
```
Comment (karlbecker) [inline src/summon/browzine-summon-adapter.js]: _non-blocking PR comment_

Could be a potential improvement in the future, but what if we collected all these default strings and put them at the top of the file in one place?  Would let John or someone else very easily see, "oh, here are all the things that a user might see in the UI," and allow easier customization that way.

File: src/summon/browzine-summon-adapter.js [hunk 6]
```diff
@@ -197,14 +246,20 @@ browzine.summon = (function() {
...
 
       if(browzineWebLink) {
-        var template = buildTemplate(scope, browzineWebLink);
+        var template = browzineWebLinkTemplate(scope, browzineWebLink);
         $(documentSummary).find(".docFooter .row:eq(0)").append(template);
       }
 
-      if(coverImageUrl && browzineEnabled) {
+      if(coverImageUrl && !defaultCoverImage) {
         $(documentSummary).find(".coverImage img").attr("src", coverImageUrl).attr("ng-src", coverImageUrl).css("box-shadow", "1px 1px 2px #ccc");
```
[the remark this answers is not in the document]
Comment (tiandavis) [reply src/summon/browzine-summon-adapter.js]: > is this an optimization to reduce the number of times we're trying to set the cover image

`browzineEnabled` stops the default cover image from appearing. But we want to expand Journal Cover coverage to show any Journal Covers we have available. The presence of the default cover image means we do not have the actual Journal Cover, hence the new `defaultCoverImage` check.

[the remark this answers is not in the document]
Comment (karlbecker) [reply src/summon/browzine-summon-adapter.js]: > But we want to expand Journal Cover coverage to show any Journal Covers we have available.

Right - in the past, we would show our default cover image in Primo/Summon, correct?  

But now, with this change, we will _not_ show our default cover image there, and we will just leave whatever Primo/Summon use by default... if I'm reading this right, anyway.

Am I grokking it? 🤔

File: tests/unit/models/primo.js (hunk 2 above)
Comment (tiandavis) [reply tests/unit/models/primo.js]: @karlbecker Fantastic question @karlbecker! TLDR; is this is how we'd want to think about Acceptance/Integration Tests, not necessarily Unit Tests. E.g. notice our [Acceptance Tests](https://github.com/thirdiron/browzine-discovery-service-adapters/blob/develop/tests/acceptance/browzine-primo-adapter-test.js#L6) have that high-level journal vs article organization. Whereas, our Unit Test here is more granularly specific to a methods and sometimes those methods test for both journal or article use. 

I could see updating the `it` statement to focus more on what the method should be doing and not so much on the API details though. Will update the same.

Good point on Cucumber though. Noting, I wouldn't knock a development team that used Cucumber for their Acceptance Tests since there is the aspect of bridging the gap between development and business requirements. But, I my mind, the biggest benefit is using it at the higher Integration Testing Environment level since these business requirements typically involve multiple applications integrating to succeed on that higher level business requirement: e.g. apps here, `browzine-discovery-service-adapters` and `browzine_cms`.

File: tests/unit/models/primo.js (hunk 3 above)
Comment (tiandavis) [reply tests/unit/models/primo.js]: Thanks Karl! See, https://github.com/thirdiron/browzine-discovery-service-adapters/pull/45#discussion_r208045501

File: README.md [hunk 7]
```diff
@@ -67,6 +67,9 @@ var browzine = {
   summonArticleWording: "View Complete Issue",
   summonJournalBrowZineWebLinkText: "Browse Now",
   summonArticleBrowZineWebLinkText: "Browse Now",
+  summonArticlePDFDownloadLinkEnabled: true,
+  summonArticlePDFDownloadWording: "Article PDF",
```
[the remark this answers is not in the document]
Comment (tiandavis) [reply README.md]: Yeah, with these configuration properties, it's like, "sigh, people are already using them, so let's jus try our best and then support until these adapters get integrated into the host product officially."

File: src/summon/browzine-summon-adapter.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (tiandavis) [reply src/summon/browzine-summon-adapter.js]: > in the past, we would show our default cover image in Primo/Summon, correct? 

No. Because the `browzineEnabled` flag seemed to always come with an actual Journal Cover.

File: tests/unit/models/primo.js (hunk 4 above)
Comment (tiandavis) [reply tests/unit/models/primo.js]: Good idea Karl. Adding these now.

**Comment by tiandavis** [comment]
Thanks for all the feedback folks!
